[Model] Bert Embedding Model

laishzh opened the pull request:
Implement Bert Embedding Model

This PR implements the Bert Embedding Model which discussed in #5179.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

robertgshaw2-neuralmagic commented on vllm/model_executor/models/bert_embedding.py:
```diff
@@ -0,0 +1,399 @@
+    ) -> Optional[PoolerOutput]:
+        return self._pooler(hidden_states, pooling_metadata)
+
+    def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
```
Can you move this function down to the bottom of the file

laishzh replied:
OK, leave a TODO here. I will move this class to the bottom later.

robertgshaw2-neuralmagic commented on vllm/model_executor/models/bert_embedding.py:
```diff
@@ -0,0 +1,399 @@
+        super().__init__()
+        self.size = config.hidden_size
+
+        self.word_embeddings = nn.Embedding(config.vocab_size,
```
Could you look into using `VocabParallelEmbedding` from our parallel layers?

Using the `nn.Embedding` will not work with tensor parallelism

laishzh replied:
OK, I noticed this feature before. Already updated.

robertgshaw2-neuralmagic commented on vllm/model_executor/models/bert_embedding.py:
```diff
@@ -0,0 +1,399 @@
+                                                  config.hidden_size)
+        self.LayerNorm = nn.LayerNorm(config.hidden_size,
+                                      eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
```
Do we need dropout for inference?

laishzh replied:
All dropouts have been removed.

robertgshaw2-neuralmagic commented on vllm/model_executor/models/bert_embedding.py:
```diff
@@ -0,0 +1,399 @@
+            bias=True,
+            quant_config=quant_config)
+
+        self.attn = Attention(
```
We need to use bidirectional attention here rather than causal attention

laishzh replied:
Do you mean to implement a new type of attention in vllm? Not sure if there is a way to use the current attention implementation with different parameters.

laishzh replied:
@robertgshaw2-neuralmagic hi, I'd like to hear your thoughts on this point. The framework of BERT model is almost completed. But the BertSelfAttention output differs from transformers. After diving into the Attention implementation, there are massive changes of Attention are needed to use bidirectional attention.
I also find some related discussions(https://github.com/vllm-project/vllm/pull/3117#issuecomment-2159049996). I think this PR is depended on #4942. What's your opinions?

robertgshaw2-neuralmagic replied:
@laishzh sorry missed this note!

Yes. Pull in the `BertSelfAttention` from that PR. #4942 should land this week

robertgshaw2-neuralmagic commented:
Thanks! You're on the right track here

mgoin commented on vllm/model_executor/models/bert_embedding.py:
```diff
@@ -0,0 +1,399 @@
+                                                config.hidden_size)
+        self.token_type_embeddings = nn.Embedding(config.type_vocab_size,
+                                                  config.hidden_size)
+        self.LayerNorm = nn.LayerNorm(config.hidden_size,
```
Follow naming standards for layernorm
```suggestion
        self.layernorm = nn.LayerNorm(config.hidden_size,
```

laishzh replied:
Followed the suggestion, and added renaming of LayerNorm parameter when loading weights.

mgoin commented on vllm/model_executor/models/bert_embedding.py:
```diff
@@ -0,0 +1,399 @@
+    def __init__(self, config: BertConfig):
+        super(BertSelfOutput, self).__init__()
+        self.dense = nn.Linear(config.hidden_size, config.hidden_size)
+        self.LayerNorm = nn.LayerNorm(config.hidden_size,
+                                      eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
```
ditto on layernorm and dropout

mgoin commented on vllm/model_executor/models/bert_embedding.py:
```diff
@@ -0,0 +1,399 @@
+    def __init__(self, config: BertConfig):
+        super().__init__()
+        self.dense = nn.Linear(config.intermediate_size, config.hidden_size)
+        self.LayerNorm = nn.LayerNorm(config.hidden_size,
+                                      eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
```
ditto on layernorm and dropout

laishzh commented:
> Thanks! You're on the right track here

Wow! Really appreciate your guidance.

laishzh commented on vllm/model_executor/models/bert_embedding.py:
```diff
@@ -0,0 +1,399 @@
+                weight_loader = param.weight_loader
+                weight_loader(param, loaded_weight, shard_id)
+                break
+            else:
```
Yes, I also felt confused at the first time. I rewote this part just now. It's supposed to be easier to understand. Have any further suggestions?

afeldman-nm committed:
wip

afeldman-nm committed:
BART almost passing profile_run()

afeldman-nm committed:
wip bart

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_cross_attn' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_model_runner_bart' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
BART passes profile run

afeldman-nm committed:
fixed prompt processing bug that was preventing inference from starting

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
wip bart-cnn summarization example

afeldman-nm committed:
fixed a number of bugs related to BART decode-phase; added support for the particular architecture alias used by bart-large-cnn

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
BART e2e test runs but does not pass

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
removed extra line

afeldman-nm committed:
changed nested if/else to elif/else in xformers mask computation code

afeldman-nm committed:
reorganized helper functions that were only being used for testing into tests/kernels/utils.py from vllm/utils.py

afeldman-nm committed:
removed attention_type

afeldman-nm committed:
typing and formatting

afeldman-nm committed:
typing and formatting; fixed escape sequences in comments

afeldman-nm committed:
moved make_tensor_with_pad() helper function back to vllm.utils

afeldman-nm committed:
formatting

afeldman-nm committed:
merge; a lot of formatting fixes to bart code but not fully passing

afeldman-nm committed:
removed unnecessary positions arguments from BART encoder, decoder forward()

afeldman-nm committed:
some reformatting

afeldman-nm committed:
fixed bug caused by overzealous refactoring

afeldman-nm committed:
BART with new explanatory comments & passing formatting tests

afeldman-nm committed:
Removed unnecessary position arguments from BART routine; formatting

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_cross_attn' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
removed redundant elif

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
reverted unnecessarily vllm/utils.py changes

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
manually merged BART code in from previous modelrunner attempt, it won't work tho until new modelrunner is finished

afeldman-nm committed:
Merge branch 'infra_enc_dec_cross_attn' into infra_enc_dec_model_runner2

afeldman-nm committed:
refactored where a number of constants are stored, primarily constants related to encoder/decoder

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_cross_attn' into infra_enc_dec_model_runner2

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Update vllm/attention/backends/torch_sdpa.py

nit: This will reduce the number of line changes and make the code look better.

Co-authored-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

afeldman-nm committed:
Update vllm/attention/layer.py

Co-authored-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_cross_attn_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_cross_attn_reviews' into infra_enc_dec_cross_attn

afeldman-nm committed:
merged in first pieces of woosuk feedback & latest main; formatting

afeldman-nm committed:
fixed specific point-changes requested by woosuk

afeldman-nm committed:
test_encoder_decoder_attn.py cleanup

afeldman-nm committed:
tests/kernels/utils.py cleanup

afeldman-nm committed:
vllm/attention/backends/abstract.py cleanup

afeldman-nm committed:
vllm/attention/backends/blocksparse_attn.py cleanup

afeldman-nm committed:
vllm/attention/backends/flash_attn.py cleanup

afeldman-nm committed:
cleaning up a number of backends & backends utils.py

afeldman-nm committed:
xformers backend cleanup

afeldman-nm committed:
formatting

afeldman-nm committed:
Merge branch 'infra_enc_dec_cross_attn' into infra_enc_dec_model_runner2

afeldman-nm committed:
merge from main

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner2

afeldman-nm committed:
wip scheduler

afeldman-nm committed:
formatting

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
scheduler supports encoder-/cross-attention & passes existing scheduler tests, but needs new encoder/decoder-specific tests

afeldman-nm committed:
scheduler test passes

afeldman-nm committed:
formatting

afeldman-nm committed:
formatting

laishzh committed:
Merge remote-tracking branch 'bert_deps/afeldman-nm/infra_enc_dec_model_runner'

afeldman-nm committed:
wip model runner

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner

afeldman-nm committed:
wip modelrunner

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
enc/dec decoder test working, sans sampling check

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
EncoderDecoderModelInput correctly handles encoder token/position fields

afeldman-nm committed:
format

afeldman-nm committed:
bart test skipped on CPU version of vllm

afeldman-nm committed:
Formatting

afeldman-nm committed:
refactored AttentionType and related imports; skip BART test definitions entirely if on vllm CPU version (to avoid xformers import

afeldman-nm committed:
formatting

afeldman-nm committed:
wip

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_model_runner_reviews' into infra_enc_dec_model_runner

afeldman-nm committed:
prompt type checks

afeldman-nm committed:
Format

afeldman-nm committed:
modified LLM.generate() error message

afeldman-nm committed:
wip engine is_encoder_decoder() setting

afeldman-nm committed:
formatting

afeldman-nm committed:
formatting?

afeldman-nm committed:
Sequence may be constructed with encoder/decoder LLMInput configurations

afeldman-nm committed:
wip but having wllm.commit_id error

afeldman-nm committed:
correctly constructing enc/dec sequences

afeldman-nm committed:
formatting

afeldman-nm committed:
wip

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
wip

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
wip

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
conftest: encoder/decoder example prompts

afeldman-nm committed:
Hfrunner greedy logprobs limit

afeldman-nm committed:
input preparation now includes encoder-oriented input setup:

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
VllmRunner encoder/decoder methods

afeldman-nm committed:
bart parallel vocab

afeldman-nm committed:
zip enc/dec prompts; formatting

afeldman-nm committed:
wip

afeldman-nm committed:
encoder/decoder input processing; formatting

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
incorporated encoder sequence into request-add functionality

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
fixed decoder-only bug

afeldman-nm committed:
bugfix

afeldman-nm committed:
wip

afeldman-nm committed:
wip

afeldman-nm committed:
fixed scheduler bug

afeldman-nm committed:
format

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
bugfix

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
fixed sequence bug

afeldman-nm committed:
added examples utils w/ context manager for backend override; applied to enc/dec example to force XFormers

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
formatting

afeldman-nm committed:
added encoder/decoder example to examples test

afeldman-nm committed:
wip refactoring

afeldman-nm committed:
refactor

afeldman-nm committed:
RequestOutput & SequenceGroup now include encoder prompt in output, as does encoder/decoder example.

afeldman-nm committed:
wip parallel bart but encountering GPU count issue

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
tweaks to enc/dec example

afeldman-nm committed:
formatting

afeldman-nm committed:
wip

afeldman-nm committed:
modified HF behavior in BART test to be truly greedy

afeldman-nm committed:
formatting

afeldman-nm committed:
wip parallelizing BART

afeldman-nm committed:
wip activation parallelization

afeldman-nm committed:
merged in upstream changes; left some formatting issues which I expect to be fixed upstream

afeldman-nm committed:
Merge branch 'infra_enc_dec_model_runner' into infra_enc_dec_model_runner_parallel_bart

afeldman-nm committed:
wip:

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_model_runner_reviews' into infra_enc_dec_model_runner_parallel_bart

laishzh committed:
Merge remote-tracking branch 'bert_deps/afeldman-nm/infra_enc_dec_model_runner'

afeldman-nm committed:
wip merge

afeldman-nm committed:
additional merge

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
CommonMetadataBuilder sets block_tables constructor arg of metadata

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_model_runner' into infra_enc_dec_model_runner_parallel_bart

afeldman-nm committed:
equalized some generation/sampling config settings between enc/dec HF,vLLM, nonetheless still not perfect match

afeldman-nm committed:
Merge branch 'infra_enc_dec_model_runner' into infra_enc_dec_model_runner_parallel_bart

afeldman-nm committed:
BART MLPs parallelized

afeldman-nm committed:
parallelized BART learned positional embedding

afeldman-nm committed:
all attention layer output linears are parallelized

afeldman-nm committed:
encoder attention & decoder self-attention parallelized

afeldman-nm committed:
parallelized LM head

afeldman-nm committed:
parallelized enc/dec cross-attention, using a slight hack

afeldman-nm committed:
fixed bug where underlying Attention was constructed using full head-count

afeldman-nm committed:
bart is parallelized, modulo an unfortunate hack for QKVParallelLinear in cross-attention

afeldman-nm committed:
commented out BART TP=4

laishzh committed:
Merge remote-tracking branch 'bert_deps/afeldman-nm/infra_enc_dec_model_runner'

afeldman-nm committed:
upstream merge

afeldman-nm committed:
Merge branch 'infra_enc_dec_model_runner_reviews' into infra_enc_dec_model_runner

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
fixed bug in how conftest was handling HF encoder/decoder outputs; disabled HF engram repeat checks

afeldman-nm committed:
set up None/empty str tests which are not passing

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'infra_enc_dec_model_runner_correctness' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
wip

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
deleted unnecessary dependency

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
empty-string decoder input is now handled for encoder/decoder

afeldman-nm committed:
enc/dec handles empty str and None decoder prompts correctly

afeldman-nm committed:
typing fix

afeldman-nm committed:
fixed bugs in handling non-text formats for individual prompts

afeldman-nm committed:
example includes prompt zipper

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
_free_seq_group() -> _free_seq_group_cross_attn_blocks()

afeldman-nm committed:
refactoring

afeldman-nm committed:
formatting

afeldman-nm committed:
removed unnecessary argument reordering

afeldman-nm committed:
enc/dec example comments'

afeldman-nm committed:
responses to feedback

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
fixed bug caused by upstream refactoring

afeldman-nm committed:
formatting

afeldman-nm committed:
upstream merge

afeldman-nm committed:
Removed lora from enc/dec model runner

afeldman-nm committed:
Merge branch 'main' into infra_enc_dec_model_runner_reviews

afeldman-nm committed:
removed lora & vision & mm code from enc/dec modelrunner

afeldman-nm committed:
checked out examples/offline_inference.py from main

afeldman-nm committed:
updated RequestOutput docstring

afeldman-nm committed:
updated RequestOutput docstring

afeldman-nm committed:
Sequence docstring

afeldman-nm committed:
removed flashinfer references from enc/dec modelrunner

afeldman-nm committed:
format

afeldman-nm committed:
removed chunked prefill logic/docstring text from enc/dec modelrunner

afeldman-nm committed:
removed prefix caching from enc/dec modelrunner

laishzh committed:
Merge remote-tracking branch 'bert_deps/afeldman-nm/infra_enc_dec_model_runner'

laishzh committed:
Merge remote-tracking branch 'origin/main'

# Conflicts:
#	.buildkite/test-pipeline.yaml
#	examples/offline_inference_encoder_decoder.py
#	tests/conftest.py
#	tests/core/test_scheduler_encoder_decoder.py
#	tests/kernels/test_encoder_decoder_attn.py
#	tests/models/test_bart.py
#	tests/worker/test_encoder_decoder_model_runner.py
#	vllm/core/scheduler.py
#	vllm/engine/llm_engine.py
#	vllm/inputs/__init__.py
#	vllm/inputs/data.py
#	vllm/model_executor/models/bart.py
#	vllm/sequence.py
#	vllm/utils.py
#	vllm/worker/enc_dec_model_runner.py
#	vllm/worker/worker.py

laishzh committed:
(WIP)feat: EmbeddingModelRunner support encoder model

laishzh committed:
WIP: bert embedding

laishzh committed:
feat: full pipeline

laishzh committed:
chore: recover

laishzh committed:
feat: default bos_token_id of encoder model

laishzh committed:
feat: recover sequence

laishzh committed:
feat: embedding model forward

laishzh committed:
chore: recover unchanged files

laishzh committed:
chore: recover

laishzh committed:
feat: fix lint

laishzh committed:
feat: fix lint

laishzh committed:
feat: fix lint

laishzh committed:
feat: embedding model prompt

laishzh committed:
feat: bos_token_id

laishzh committed:
fix: fix hint

laishzh committed:
feat: remove embedding block space manager

laishzh committed:
feat: enc_dec_runner base

laishzh committed:
Merge branch 'main' into main

laishzh committed:
feat: fix lint

laishzh committed:
feat: model input

laishzh committed:
chore: fix lint

laishzh committed:
feat: move BertEmbeddingModel to the end of file

laishzh committed:
feat: remove embedding_model_block_manager.py

laishzh committed:
chore: fix lint

laishzh committed:
feat: modify test_embedding

laishzh committed:
feat: bert embedding implemented, but still have some bugs with mistral,

laishzh committed:
feat: some changes on test_embedding.py

laishzh committed:
Merge branch 'main' of https://github.com/vllm-project/vllm

# Conflicts:
#	vllm/core/embedding_model_block_manager.py

laishzh committed:
feat: fix lint

laishzh committed:
feat: fix lint

laishzh commented:
@robertgshaw2-neuralmagic @mgoin Hi, just update this work. 
Because the BertEmbeddingModel is encoder-only architecture. I refactored the EmbeddingModelRunner as child class of EncoderDecoderModelRunner to reuse the encoder-decoder framework.

laishzh commented on vllm/worker/embedding_model_runner.py:
```diff
@@ -137,14 +149,34 @@ def prepare_model_input(
         # Prepare PoolingMetadata.
-        assert model_input.seq_lens is not None
+        seq_lens = model_input.seq_lens\
+            if not self.model_config.is_encoder_model \
```
@robertgshaw2-neuralmagic @mgoin Here still have a question. How to determine precisely which model is encoder model or decoder model. As an example, there is lake of fields `is_decoder_model` or `is_encoder_model` in `config.json` of `Mistral Model` which is implemented before. Here is the link: [config.json of Mistral](https://huggingface.co/intfloat/e5-mistral-7b-instruct/blob/main/config.json).

maxdebayser committed:
Merge branch 'main' into bert

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>

maxdebayser commented:
Hi @laishzh, I'm working on another PR that is based on yours. In case it helps, I've solved the latest merge conflicts of your branch with `main` here: https://github.com/maxdebayser/vllm/tree/bert

laishzh committed:
Merge remote-tracking branch 'origin/main'

# Conflicts:
#	vllm/inputs/data.py

maxdebayser committed:
Merge branch 'upstream_main' into bert

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into bert

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
@laishzh @robertgshaw2-neuralmagic , I've solved that lastest merge conflicts here: https://github.com/maxdebayser/vllm/tree/bert

laishzh committed:
Merge remote-tracking branch 'maxdebayser/bert'

laishzh committed:
feat: revert embedding_block_manager

laishzh committed:
Merge branch 'origin/main'

laishzh committed:
feat: update with origin/main

laishzh commented:
@robertgshaw2-neuralmagic Please draw your attention. I just revert the change to EmbeddingModelBlockManager. And there is still leaving a problem of how to distinguish the model arch(Encoder-Only, or Decoder-Only model), which is to determine the seq_len when pooling(https://github.com/vllm-project/vllm/pull/5447#discussion_r1750487428). Please let me know if any changes are needed.

robertgshaw2-neuralmagic commented:
Hey @laishzh - there were a few issues with this PR. Specifically, it uses the Encoder-Decoder pathway and the BertModel is not implemented properly (loading is not canonical, tensor parallelism does not work, and the pooling logic is not correct)

I need get this landed ASAP, so I finished off the PR here. https://github.com/vllm-project/vllm/pull/9056

I will added you as a co-author.

laishzh commented:
> Hey @laishzh - there were a few issues with this PR. Specifically, it uses the Encoder-Decoder pathway and the BertModel is not implemented properly (loading is not canonical, tensor parallelism does not work, and the pooling logic is not correct)
> 
> I need get this landed ASAP, so I finished off the PR here. #9056
> 
> I will added you as a co-author.

I'm OK. Very glad to see this feature could be supported in vllm. I will delve into those points that you mentioned. Thanks for your work!

DarkLight1337 commented:
Closing as superseded by #9056
